LinqLesson/LinqLesson/Program.cs

```diff
@@ -0,0 +1,336 @@
+
+        Console.WriteLine($"Most used tag: {mostUsedTag.Tag} ({mostUsedTag.Count})");
+    }
+
+    public static void MinMaxDistanceBetweenPersons(this IEnumerable<Person> persons)
+    {
+        var distances = persons.Join(persons,
+            person => true,
+            person => true,
+            (person1, person2) => new
+            {
+                First = person1,
+                Second = person2
+            })
+            .Where(twoPersons => twoPersons.First != twoPersons.Second)
+            .Select(twoPersons =>
+            {
+                var distance = Math.Sqrt(Math.Pow(twoPersons.First.Latitude - twoPersons.Second.Latitude, 2)
+                    + Math.Pow(twoPersons.First.Longitude - twoPersons.Second.Longitude, 2));
+
+                return new
+                {
+                    First = twoPersons.First,
+                    Second = twoPersons.Second,
+                    Distance = distance
+                };
+            })
+            .ToArray();
+
+        var min = distances.MinBy(x => x.Distance);
+        var max = distances.MaxBy(x => x.Distance);
+
+        Console.WriteLine($"Min distance is between {min.First} and {min.Second} is {min.Distance}");
+        Console.WriteLine($"Max distance is between {max.First} and {max.Second} is {max.Distance}");
+    }
+
+    public static void GroupByEyeColor(this IEnumerable<Person> persons)
+    {
+        var result = persons.GroupBy(person => person.EyeColor)
+            .Select(group => new
+            {
+                EyeColor = group.Key,
+                Count = group.Count()
+            });
+
+        foreach (var item in result)
+        {
+            Console.WriteLine($"{item.Count} has {item.EyeColor} eye color");
+        }
+    }
+
+    public static void CollectAllEmail(this IEnumerable<Person> persons)
+    {
+        var emails = persons.Select(person => person.Email)
+            .Aggregate(new StringBuilder(),
+                (sb, email) => sb.Append(email).Append(";"),
+                sb => sb.ToString());
+
+        Console.WriteLine($"All emails: {emails}");
+    }
+
+    public static void MaxFriends(this IEnumerable<Person> persons, int top)
+    {
+        var topFriends = persons.OrderByDescending(person => person.Friends.Length)
+            .Take(top);
+
+        var i = 0;
+        foreach (var topFriend in topFriends)
+        {
+            Console.WriteLine($"Top friend {++i}: {topFriend.Name} with {topFriend.Friends.Length} friends");
+        }
+    }
+
+    public static void EverybodyHasEmail(this IEnumerable<Person> persons)
+    {
+        var allHasEmails = persons.All(person => !string.IsNullOrEmpty(person.Email));
+        Console.WriteLine($"Everybody has email: {allHasEmails}");
+    }
+
+    #region Homework
+    public static void TheMostFarthestPerson(this IEnumerable<Person> persons)
+    {
+        var northernPerson = persons.OrderByDescending(x => x.Latitude).First();
```

Comment:
it's better to use here max/min because it's easier to found max/min then to sort